### PR TITLE
#43 TkGitHubAuthorization test

### PR DESCRIPTION
**Pull request by olenagerasimova**
For #43 
- refactored TkGitHubAuthorization to make testing possible
- implemented unit test
- removed puzzle

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#59](https://github.com/g4s8/ghman/issues/59) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#59](https://github.com/g4s8/ghman/issues/59) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CJT4R32Q3/d78e7687-b3c1-4530-b52a-91eb8909db18); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/com/g4s8/ghman/user/User.java [hunk 1]
```diff
@@ -71,12 +72,27 @@
...
+        public Fake(final Github github, final AtomicInteger auth) {
             this.github = github;
+            this.auth = auth;
+        }
+
+        /**
+         * Ctor.
+         * @param github Github api
+         */
+        public Fake(final Github github) {
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/user/User.java]: @olenagerasimova the secondary constructor must be declared before the primary

File: src/main/java/com/g4s8/ghman/user/Users.java [hunk 2]
```diff
@@ -55,31 +56,39 @@
...
+         * Users.
          */
-        private final Github github;
+        private final Iterable<User> users;
+
+        /**
+         * Ctor.
+         * @param users Users
+         */
+        public Fake(final Iterable<User> users) {
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/user/Users.java]: @olenagerasimova the secondary constructor must be declared before the primary

File: src/main/java/com/g4s8/ghman/web/TkApp.java [hunk 3]
```diff
@@ -54,7 +56,9 @@
                         new TkFork(
                             new FkParams(
                                 "code", ".+",
-                                new TkGitHubAuthorization(data)
+                                new TkGitHubAuthorization(
+                                    new PgUsers(data), new EnvironmentVariables()
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/web/TkApp.java]: @olenagerasimova `EnvironmentVariables` must be declared in the constructor to of `TkApp` and passed to all the objects that needs it

File: src/main/java/com/g4s8/ghman/web/TkGitHubAuthorization.java [hunk 4]
```diff
@@ -36,43 +38,65 @@
...
+        this.users = users;
+        this.auth = auth;
+    }
+
+    /**
+     * Ctor.
+     * @param users Users
+     * @param env Environment
+     */
+    TkGitHubAuthorization(final Users users, final EnvironmentVariables env) {
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/web/TkGitHubAuthorization.java]: @olenagerasimova the secondary constructor must be declared before the primary

File: src/main/java/com/g4s8/ghman/web/TkGitHubAuthorization.java [hunk 5]
```diff
@@ -36,43 +38,65 @@
...
     /**
      * Data source.
      */
-    private final DataSource data;
+    private final Users users;
+
+    /**
+     * Authorization function.
+     */
+    private final Func<org.takes.Request, String> auth;
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/web/TkGitHubAuthorization.java]: @olenagerasimova introduce a meaningful abstraction for this. I think the current implementation in the secondary constructor should reside in its own class and there should be a `Fake` implementation for use in the tests.

File: src/test/java/com/g4s8/ghman/web/TkGitHubAuthorizationTest.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    @Test
+    void returnsMessage() throws IOException {
+        new Assertion<>(
+            "Returns user friendly message",
+            new RsPrint(
+                new TkGitHubAuthorization(
+                    new Users.Fake(new IterableOf<>(new User.Fake(new MkGithub()))),
+                    req -> "abc"
+                ).act(new RqWithAuth(new Identity.Simple("urn:fake:5"), new RqFake()))
+            ).print(),
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/web/TkGitHubAuthorizationTest.java]: @olenagerasimova I think it is safer/stricter to use `printBody` here (but I may be mistaken)

**Comment by olenagerasimova** [comment]
@victornoel done

**Comment by victornoel** [comment]
@olenagerasimova thank you. You forgot https://github.com/g4s8/ghman/pull/59#discussion_r300837451 though.

**Comment by olenagerasimova** [comment]
@victornoel sorry, fixed missed comment

**Comment by victornoel** [comment]
@g4s8 should be good to merge

**Comment by 0crat** [comment]
@victornoel[/z](https://www.0crat.com/u/victornoel) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by victornoel** [comment]
@g4s8 ping

**Comment by victornoel** [comment]
@olenagerasimova can you rebase your branch onto master instead of merging master into your branch? This makes git history ugly to look at AND PR harder to review because the diff is polluted by changes coming from master.

**Comment by olenagerasimova** [comment]
@victornoel ok

**Comment by victornoel** [comment]
> @olenagerasimova can you rebase your branch onto master instead of merging master into your branch? This makes git history ugly to look at AND PR harder to review because the diff is polluted by changes coming from master.

@olenagerasimova actually after discussion in #65, I think we are going toward merging master into branches and not using rebase, so you can just ignore the comment above.

**Comment by victornoel** [comment]
@olenagerasimova the branch is in conflict with master

**Comment by olenagerasimova** [comment]
@victornoel @g4s8 conflicts resolved

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
Code review was too long (13 days), architects (@g4s8) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
Job [#59](https://github.com/g4s8/ghman/issues/59) is not in the agenda of @victornoel[/z](https://www.0crat.com/u/victornoel), can't inspect

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#59](https://github.com/g4s8/ghman/issues/59) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ammaratef45** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)

**Pull request merged**